### PR TITLE
hil: Use multi-part download for hil

### PR DESCRIPTION
**Pull request by AlexKaravaev**
AWS rust sdk is slow compared to cli, multipart download can speedup the process significantly,
see https://github.com/awslabs/aws-sdk-rust/discussions/1024. 

The logic for the change is mostly adapted from: https://github.com/smithy-lang/smithy-rs/blob/main/aws/sdk/benchmarks/s3-throughput/benchmark/src/multipart_get.rs.

For me the speedup was around 2x in Munich(It should be different for different latencies).

To be noted, that requests to s3 do cost something, but as I understood it's insignificant amount: https://aws.amazon.com/s3/pricing/?p=pm&c=s3&z=4. Lmk if I missed anything.

0.0004 $ per 1000 requests and with multi-part download we have ~40 requests per GB.

### CODE REVIEW

**Review by TheButlah — CHANGES_REQUESTED** [review]
Awesome! I didn't know about the multi-part downloads, TIL.

Excited to land this :)

File: hil/src/download_s3.rs [hunk 1]
```diff
@@ -1,126 +1,220 @@
...
     /// If a file exists, abort
     Abort,
 }
 
-/// `out_path` is the final path of the file after downloading.
+#[derive(Debug)]
+struct ContentRange {
+    start: u64,
+    end: u64,
+}
```
[the remark this answers is not in the document]
Comment (AlexKaravaev) [reply hil/src/download_s3.rs:41]: Everything in your comments makes sense. I rather lazely took the linked in the description upstream aws sdk code and just made it work. Will address the comments.

File: hil/src/download_s3.rs [hunk 2]
```diff
@@ -113,58 +114,62 @@ pub async fn download_url(
...
-                        bytes_to_download >> 20,
-                        pct,
-                    );
+            loop {
+                let range_option = {
+                    let mut ranges_lock = ranges.lock().await;
+                    ranges_lock.next()
+                };
+
+                let range = match range_option {
```
Comment (TheButlah) [inline hil/src/download_s3.rs:139]: prefer let .. else syntax for simple pattern matching where the failure case evaluates to `!` (such as when doing `return` or `break`).

**Review by TheButlah — CHANGES_REQUESTED** [review]
1 tiny remaining change (the let .. else) and it will be good to go.

**Pull request merged**